### PR TITLE
Remove obsolete workarounds for pre-C++17

### LINKED ISSUE #1495: Switch to C++17 `constexpr`
In `meta/CudaVectorArrayWrapper.hpp` we find the following section:

```c++
#    if(__cplusplus > 201402L)
#        define CXX17_CONSTEXPR constexpr
#    else
// This is mostly needed because of nvcc 9 and should be removed upon moving to C++17.
// Maybe some others need it too. GNU STL library uses the same trick.
#        define CXX17_CONSTEXPR
#    endif
```

All prerequisites are met to remove this.

### PR DESCRIPTION
**Pull request by bernhardmgruber**
* switch clang-7/nvcc to clang-8/nvcc
* Remove obsolete workarounds for pre-C++17

### CODE REVIEW

**Comment by bernhardmgruber** [comment]
Looks like we missed something with the C++17 upgrade:
linux_nvcc-11.0_clang-6_release_cuda_only:
```
nvcc warning : The -std=c++17 flag is not supported with the configured host compiler. Flag will be ignored.
/home/runner/work/alpaka/alpaka/include/alpaka/kernel/Traits.hpp(216): error #135: namespace "std" has no member "invoke_result_t"
```

File: cmake/alpakaCommon.cmake [hunk 1]
```diff
@@ -407,6 +407,15 @@ if(ALPAKA_ACC_GPU_CUDA_ENABLE)
...
 
+            # If host compiler is clang < 8, nvcc does not forward -std=c++17
+            if (CMAKE_CUDA_HOST_COMPILER MATCHES ".*clang\\+\\+.*")
+                execute_process(COMMAND ${CMAKE_CUDA_HOST_COMPILER} --version OUTPUT_VARIABLE CLANG_VERSION_STRING)
+                string(REGEX MATCH "clang version ([0-9]+)\\.[0-9]+\\.[0-9]+" DUMMY ${CLANG_VERSION_STRING})
+                if (CMAKE_MATCH_1 LESS 8)
+                    alpaka_set_compiler_options(DEVICE target alpaka $<$<COMPILE_LANGUAGE:CUDA>:-Xcompiler=-std=c++17>)
+                endif ()
+            endif ()
+
```
Comment (bernhardmgruber) [inline cmake/alpakaCommon.cmake]: Unfortunately, CMake does not detect and provide better variables to inspect the CUDA host compiler.

**Review by sbastrakov — DISMISSED** [review]

**Review by psychocoderHPC — DISMISSED** [review]

**Comment by bernhardmgruber** [comment]
So my CMake workaround seems to correctly detect the host compiler and cause this invocation:
```
cd /home/runner/work/alpaka/alpaka/build/example/bufferCopy && /usr/local/cuda-11.0/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/clang++ -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_BLOCK_SHARED_DYN_MEMBER_ALLOC_KIB=47 -DALPAKA_CI -DALPAKA_DEBUG=0 -DALPAKA_DEBUG_OFFLOAD_ASSUME_HOST -DALPAKA_OFFLOAD_MAX_BLOCK_SIZE=256 -I/home/runner/work/alpaka/alpaka/include -isystem=/usr/local/cuda-11.0/targets/x86_64-linux/include -isystem=/home/runner/work/alpaka/alpaka/boost -isystem=/usr/local/cuda-11.0/include -O3 -DNDEBUG --generate-code=arch=compute_52,code=[compute_52,sm_52] -Xcompiler=-std=c++17 --extended-lambda --expt-relaxed-constexpr -Xcudafe=--display_error_number -Xcudafe=--diag_suppress=esa_on_defaulted_function_ignored -Xcompiler -pthread -std=c++17 -x cu -c /home/runner/work/alpaka/alpaka/example/bufferCopy/src/bufferCopy.cpp -o CMakeFiles/bufferCopy.dir/src/bufferCopy.cpp.o
nvcc warning : The -std=c++17 flag is not supported with the configured host compiler. Flag will be ignored.
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/bits/stl_pair.h(79): error #325: inline specifier allowed on function declarations only
```
We see that `-Xcompiler=-std=c++17` and `-std=c++17` are passed to nvcc. Nvcc then claims that the host compiler does not support C++17 (which is wrong) and seems to ignore both flags.

**Comment by bernhardmgruber** [comment]
Rebased on develop, which dropped the failing runs.

**Comment by bernhardmgruber** [comment]
It seems I forgot to also upgrade the clang-7/nvcc job in GitLab in #1506. I moved it to clang-8/nvcc.

**Review by j-stephan — APPROVED** [review]

**Comment by j-stephan** [comment]
The only failing check is the CUDA 9.2 download error. Merging.

**Pull request merged**